samples/nrf_desktop/configuration/nrf52_pca20037/buttons_def.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include "buttons.h"
+
+/* This file must be included only once */
```

Comment:
This is not OK.

Please place col and row in a C file, and have users reference them by having them include a header file that declares them (with extern).

Reply:
Pros and cos were considered and this layout was selected for the project. I appreciate your opinion but I don't agree here.
If you feel this is not OK please file a bug under DESK and we can discuss it there. This is not the right place.

Reply:
We could possibly remove this problem by... the usage of Xmacros <-:[

Reply:
I assume there is a good reason to go against the norm of how C sources and header files are used.

Please document this somewhere and a reference to it from here.

Reply:
Hi @SebastianBoe , sure. This is a good idea.

@jhn-nordic , @Qbicz , @MarekPieta This may be a good chance to start creating documentation for the project configuration. At the same time we could check how intuitive is this part is. Created https://projecttools.nordicsemi.no/jira/browse/DESK-535 to handle this.

Reply:
Please rephrase this:

> This file must be included only once

which sounds like requirements documentation for something the user must take care of. To something like:

This struct enforces that this header file is only included once in the build, as including it twice would trigger a multiple definition error at link time.

Reply:
With this mechanism, what happens when you have two C files that need to access the information in row?

Reply:
@SebastianBoe , this is configuration file specific to a single module. The current architecture of the app will make it work as we have one file per module (this is a design choice). It is in the form of a header file is to allow optimizations of some loops in modules. We could change the module source to work around it but we would have to pay with additional maintenance costs. Using header file gives us the ability to keep the configuration in separate file, keep the maintenance costs intact and being able to optimize for various cases. The cons is that it is non standard way header files are used and this is the price we decided to pay. I think you are right we should add some comment about that in the header to make sure it is well understood that these files are meant for one module only.

Reply:
This is OK. I suspect that eventually, the limitation of the configuration only being accessible from a single module will cause a problem.

But until it does, I see no harm in this as long it is clearly documented what is going on here for future reviewers.

Reply:
Thanks

Reply:
Comments added.

Reply:
Fixed